### PR TITLE
Add TermReference, TypeReference, ConstructorReference

### LINKED ISSUE #2620: Proposal: some additional type safety / information around references/referents
@ChrisPenner @aryairani and I concluded something like this would probably be worthwhile to introduce, for a little bit of type safety and grokkability:

```haskell
-- Could be made newtypes, if that would be useful
type TermReference = Reference
type TypeReference = Reference

data ConstructorReference = ConstructorReference TypeReference ConstructorId
```

### PR DESCRIPTION
**Pull request by mitchellwrosen**
## Overview

This PR adds a `ConstructorReference` type, which replaces `(Reference, Int)`.

```haskell
data ConstructorReference = ConstructorReference TypeReference Int
```

It also adds two type aliases for references known to be terms/types.

```haskell
type TermReference = Reference
type TypeReference = Reference
```

Fixes #2620 

## Loose ends

No equivalent type was added to the `codebase2/` nor `hashing/` flavors of all the types yet.

### CODE REVIEW

File: parser-typechecker/src/Unison/Server/Backend.hs
Comment (mitchellwrosen) [inline parser-typechecker/src/Unison/Server/Backend.hs:625]: Unused code

**Comment by mitchellwrosen** [comment]
:thinking: @dolio In an attempt to recover a little bit of performance lost by indirection I made the fields of `ConstructorReference` strict:

```haskell
data ConstructorReference = ConstructorReference !TypeReference !ConstructorId
```

I think this is generally a good idea unless we have a specific reason to use this thing lazily. ...but we do, in the runtime!

```haskell
compile :: Var v => DataSpec -> Ctx v -> PatternMatrix v -> Term v
compile _ _ (PM []) = placeholder () "pattern match failure"
compile spec ctx m@(PM (r:rs))
  | rowIrrefutable r
  = case guard r of
      Nothing -> body r
      Just g -> iff mempty g (body r) $ compile spec ctx (PM rs)
  | PData rf <- ty
  , rf == Rf.listRef
  = match () (var () v)
      $ buildCaseBuiltin spec ctx
     <$> splitMatrixSeq (Map.keysSet ctx <> usedVars m) v m
  | PData rf <- ty
  , rf `member` builtinCase
  = match () (var () v)
      $ buildCaseBuiltin spec ctx
     <$> splitMatrixBuiltin v m
  | PData rf <- ty
  = case lookupData rf spec of
      Right cons ->
        match () (var () v)
          $ buildCase spec rf False cons ctx
         <$> splitMatrix v rf (numberCons cons) m
      Left err -> internalBug err
  | PReq rfs <- ty
  = match () (var () v) $
      [ buildCasePure spec ctx tup
      | tup <- splitMatrix v undefined [(-1,1)] m
      ] ++
      [ buildCase spec rf True cons ctx tup
      | rf <- Set.toList rfs
      , Right cons <- [lookupAbil rf spec]
      , tup <- splitMatrix v rf (numberCons cons) m
      ]
  | Unknown <- ty
  = internalBug "unknown pattern compilation type"
  where
  v = choose heuristics m
  ty = Map.findWithDefault Unknown v ctx
```

That `undefined` (type `Reference`) is now getting forced because `splitMatrix` uses it to make a `ConstructorReference`. I could remove the bangs from `ConstructorReference`, but do you know what's going on here?

**Comment by mitchellwrosen** [comment]
@dolio I've worked around the test failure with [this commit](https://github.com/unisonweb/unison/pull/2621/commits/9a12aafb626dc70e6c79ec08a402367c943c9e6e), which reverts to the comparing of `Int` before `Reference`, since the derived `Eq` instance of `ConstructorReference`, combined with a bang pattern on the `Reference` contained within, caused a regression due to that `undefined`. So no worries if you don't have time to peek at this.

**Comment by dolio** [comment]
Sorry about the delay.

I think there's nothing deep about the `undefined`. I guess I just happened to do it that way instead of sending in a `Maybe` that should always be `Just` except for `EffectPure`. If you want to make it `Maybe` instead, that might be better.

File: parser-typechecker/src/Unison/Builtin/Terms.hs
Comment (aryairani) [inline parser-typechecker/src/Unison/Builtin/Terms.hs:23]: WDYT about leaving the signature of the smart constructor unchanged, and have it construct a `ConstructorReference` internally, to reduce line noise?  Does that defeat the whole purpose?

File: parser-typechecker/src/Unison/Codebase.hs
Comment (aryairani) [inline parser-typechecker/src/Unison/Codebase.hs:202]: Maybe a `ConstructorReferenceId :: Reference.Id -> ConstructorId -> ConstructorReference` pattern synonym? Although I see `type ConstructorReferenceId` exists and so maybe that would be confusing for the pattern to not match the type.

File: parser-typechecker/src/Unison/Codebase/Branch.hs
Comment (aryairani) [inline parser-typechecker/src/Unison/Codebase/Branch.hs:137]: Tangent: Maybe we need a better name for Referent too, but I don't know what the current one means or what it should be changed to.

File: parser-typechecker/src/Unison/Hashing/V2/Convert.hs
Comment (aryairani) [inline parser-typechecker/src/Unison/Hashing/V2/Convert.hs:61]: For changes to `Term` and anything adjacent to the runtime, Is there some way to do this without additional unboxing overhead? Does it matter?

File: parser-typechecker/src/Unison/Hashing/V2/Referent.hs
Comment (aryairani) [inline parser-typechecker/src/Unison/Hashing/V2/Referent.hs:31]: I think eventually `Unison.Hashing.V2` could be its own package (`Unison.Hashing.V2.Convert` would be outside of that package), and then we'll want to decide what to do about the `Unison.Referent'` and `Unison.ConstructorReference`.

File: parser-typechecker/src/Unison/Parser.hs
Comment (aryairani) [inline parser-typechecker/src/Unison/Parser.hs:111]: Great!

File: parser-typechecker/src/Unison/Runtime/Pattern.hs
Comment (aryairani) [inline parser-typechecker/src/Unison/Runtime/Pattern.hs:157]: Why are these `Maybe`?

File: parser-typechecker/src/Unison/Runtime/Pattern.hs
Comment (aryairani) [inline parser-typechecker/src/Unison/Runtime/Pattern.hs:613]: oic.

@dolio Could you review these changes to `Unison/Runtime`?

And unless @dolio objects 😬 , let's request his review on any proposed changes to `Unison/Runtime`. @mitchellwrosen

File: parser-typechecker/src/Unison/Server/Doc.hs
Comment (aryairani) [inline parser-typechecker/src/Unison/Server/Doc.hs:278]: Is this (and line 235) longer and less readable than the non-lens version? 😅

File: parser-typechecker/src/Unison/TermParser.hs
Comment (aryairani) [inline parser-typechecker/src/Unison/TermParser.hs:206]: Maybe a pattern synonym here too, wdyt?

File: unison-cli/src/Unison/Codebase/Editor/HandleInput.hs
Comment (aryairani) [inline unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:1457]: Maybe here too, preserve the type of `Term.Constructor'`, or add another version, for convenience? Or is that counterproductive?

**Review by aryairani — APPROVED** [review]
I left some comments / ideas that would also reduce the diff, hopefully minimize merge conflicts, since it does touch a lot of places.

**Review by aryairani — COMMENTED** [review]
~~Oh but let's get Dan's approval too, before merging.~~ Sorry, just getting caught up.

File: parser-typechecker/src/Unison/Runtime/Pattern.hs
Comment (mitchellwrosen) [reply parser-typechecker/src/Unison/Runtime/Pattern.hs:157]: 🤔 Not totally sure why this is showing up in this diff. The changes were merged into trunk in #2658, so maybe I just need to merge trunk in?

File: parser-typechecker/src/Unison/Runtime/Pattern.hs
Comment (aryairani) [reply parser-typechecker/src/Unison/Runtime/Pattern.hs:613]: Oops, you already did, I missed the preceding conversations. Sorry!

File: parser-typechecker/src/Unison/Codebase/Branch.hs
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase/Branch.hs:220]: Things like this are now so much more readable 👍🏼

File: parser-typechecker/src/Unison/Server/Doc.hs
Comment (ChrisPenner) [reply parser-typechecker/src/Unison/Server/Doc.hs:278]: I think the trade off is worth it TBH; 
we should try to reduce cases where we strip the references off of things like this anyways since it's losing any type-info about the references.

File: parser-typechecker/src/Unison/TermParser.hs
Comment (ChrisPenner) [reply parser-typechecker/src/Unison/TermParser.hs:206]: Maybe the `Pattern.Constructor` can build the constructor reference from its components? 

But also notice that the ConstructorReference actually makes usages clearer later in this file where we can pass the token payload directly without unpacking.

File: unison-core/src/Unison/ConstructorReference.hs
Comment (ChrisPenner) [inline unison-core/src/Unison/ConstructorReference.hs:26]: The other modules just call this `Id` and expect qualified usage, which I think makes sense. What do you think of doing that here?

File: unison-core/src/Unison/Referent'.hs
Comment (ChrisPenner) [inline unison-core/src/Unison/Referent'.hs:39]: What's the deal with the Constructor Type here?
Seems kinda strange that Constructor ID gets folded in but the Constructor Type doesn't; however I don't think we always have the Constructor Type available do we? 

Ideally I think ConstructorType would end up in the GConstructorReference too (or in the TypeReference), but I don't know that it's possible ATM.

Not worth holding up this PR for though 👍🏼

**Review by ChrisPenner — APPROVED** [review]
Looks great! Let's get this merged quick before the conflicts pile up 🙌🏼

File: parser-typechecker/src/Unison/Builtin/Terms.hs
Comment (mitchellwrosen) [reply parser-typechecker/src/Unison/Builtin/Terms.hs:23]: I think that would be going in the opposite direction of this PR, which is to give a name to `(Reference, Int)` rather than pass them around as separate arguments everywhere.

But I think we could go further here, and have a top-level 

```haskell
Decls.isPropagatedConstructor = 
  ConstructorReference Decls.isPropagatedRef Decls.isPropagatedConstructorId
```

for this builtin, and all others, whose refs/conIds we refer to explicitly

File: parser-typechecker/src/Unison/Hashing/V2/Convert.hs
Comment (mitchellwrosen) [reply parser-typechecker/src/Unison/Hashing/V2/Convert.hs:61]: There is not any way to recover the performance characteristics of manually unpacking the `ConstructorReference` into the data structures it's contained in, no. But I don't think it matters.

File: parser-typechecker/src/Unison/Hashing/V2/Referent.hs
Comment (mitchellwrosen) [reply parser-typechecker/src/Unison/Hashing/V2/Referent.hs:31]: Yep, totally. Hashing should not pull in `Unison.ConstructorReference`, but I figured in the meantime it's ok to not define a `Hashing.ConstructorReference` right here and now because hashing already tugs on `Unison.Referent'`.

File: parser-typechecker/src/Unison/TermParser.hs
Comment (mitchellwrosen) [reply parser-typechecker/src/Unison/TermParser.hs:206]: As mentioned in another comment, one thing we could do is define a top-level name for the `()` constructor of the `()` type.

Right now it only has a top-level name for the type (`DD.unitRef`), and we pass around the `0` with it manually to refer to the one constructor.

File: unison-core/src/Unison/ConstructorReference.hs
Comment (mitchellwrosen) [reply unison-core/src/Unison/ConstructorReference.hs:26]: I think that's a reasonable idea for consistency, but what to do about these?

```haskell
-- in Unison.Reference
type TypeReferenceId = Id
type TermReferenceId = Id
```

It'd be a bigger change to go the other direction and not name things e.g. `Reference.Id`, `Referent.Id`, but that's my preference, because (for one thing) haddocks become a lot more readable.

File: parser-typechecker/src/Unison/Codebase.hs
Comment (mitchellwrosen) [reply parser-typechecker/src/Unison/Codebase.hs:202]: :man_shrugging: It's not super clear to me if we want that pattern synonym or not.

It also reminds me of the builtin suggestion from a week back:

```haskell
data Builtin a b
  = Builtin a 
  | UserDefined b
```

because what this code is really trying to do is switch on whether or not the thing's a builtin!

(...whoa, why do we have this if it error on builtins? we should just take a `ConstructorReferenceId` instead...)

File: unison-core/src/Unison/ConstructorReference.hs
Comment (ChrisPenner) [reply unison-core/src/Unison/ConstructorReference.hs:26]: Good question, I wonder if eventually it might make sense to merge everything into the same module if we can get good names for each concept, then you can just import the whole thing qualified and use `R.TypeReferenceId` etc.

For now whatever you decide is fine with me 👍🏼

**Pull request merged**
